### PR TITLE
changefeedccl: add observabilty to row fetcher rows 

### PR DESCRIPTION
**Pull request by jayshrivastava**
### sql/row: add util.EveryN arg to row fetcher for TraceKV

This change adds a `util.EveryN` parameter to the `row.Fetcher`
called `TraceKVEveryN`. When KV tracing is enabled, the row fetcher
will print ALL KVs that it sees. This will could result in millions
of rows being logged in certain cases. This change adds an `EveryN`
to limit the log frequency. `util.EveryN` is used instead of `log.EveryN`
because `log.EveryN` is a noop when verbosity is 2 or above. The row
fetcher uses verbosity level 2 to log KVs.

Release note: None
Epic: None

### changefeedccl: add observabilty to row fetcher rows
This change adds some flags which can be enabled to make
row fetchers log the KVs they are fetching. The setting
`changefeed.cdcevent.trace_kv` is passed down to the row
fetcher to enable logging. The setting
`changefeed.cdcevent.trace_kv.log_frequency` is added as
well to limit the rate of logging.

Release note: None
Epic: None

### CODE REVIEW

**Review by yuzefovich — COMMENTED** [review]
> Reviewed 1 of 1 files at r1, 4 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/112342)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jayshrivastava, @mgartner, and @miretskiy)

___
*[`pkg/ccl/changefeedccl/cdcevent/settings.go` line 29 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/112342#-NgfCNyb95OCZMAcwNbb:-NgfCNyb95OCZMAcwNbc:b-q22yj) ([raw file](https://github.com/cockroachdb/cockroach/blob/903a6d5b284426bf343e3a5eb7958e04d769557c/pkg/ccl/changefeedccl/cdcevent/settings.go#L29)):*
> ```Go
> // RowFetcherTraceKVLogFrequency controls how frequently KVs are logged when
> // KV tracing is enabled.
> var RowFetcherTraceKVLogFrequency = settings.RegisterDurationSetting(
> ```

I see, so we do have a setting for this. I think we need to lift it up into `row` package, and make `0s` the default that means that every occurrence is sampled.
___
*[`pkg/sql/row/fetcher.go` line 1144 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/112342#-NgfBY0z4V5uF6fvQNcR:-NgfBY0z4V5uF6fvQNcS:b-9fsglr) ([raw file](https://github.com/cockroachdb/cockroach/blob/615ea972162024023d2d2dc7e74d63ed66621a1f/pkg/sql/row/fetcher.go#L1144)):*
> ```Go
> 		// NB: TraceKVEvery is a util.EveryN and not a log.EveryN because
> 		// log.EveryN will always print under verbosity level 2.
> 		if rf.args.TraceKV && rf.args.TraceKVEvery.ShouldProcess(timeutil.Now()) {
> ```

I think that we need to preserve the default behavior - namely that every KV gets "fetched ..." line printed in the trace by default. We should introduce a cluster setting that would set the "KV sample frequency", which would have 0s by default (meaning everything is logged), and then we could reduce the frequency for the use case you have in mind.

**Review by miretskiy — APPROVED** [review]
:lgtm: 

but +1 to @yuzefovich comments.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/112342#-:-Nh793VL7Z5YrGnvjUda:b-xke28n)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jayshrivastava and @mgartner)

___
*[`pkg/ccl/changefeedccl/cdcevent/settings.go` line 19 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/112342#-Nh78sd8AQrCYNqnNXGv:-Nh78sd98kBpN3rf034H:b3dkehb) ([raw file](https://github.com/cockroachdb/cockroach/blob/903a6d5b284426bf343e3a5eb7958e04d769557c/pkg/ccl/changefeedccl/cdcevent/settings.go#L19)):*
> ```Go
> // RowFetcherTraceKVs controls if we write the experimental new format BACKUP
> // metadata file.
> var RowFetcherTraceKVs = settings.RegisterBoolSetting(
> ```

we probably don't need to introduce top level settings file and the settings themselves do not need to be exported.
This can be defined in the same file that uses these settings.

**Review by jayshrivastava — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/112342)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner, @miretskiy, and @yuzefovich)

___
*[`pkg/ccl/changefeedccl/cdcevent/settings.go` line 19 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/112342#-Nh78sd8AQrCYNqnNXGv:-Nh7GQTo7oIyu_g3ApfE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/903a6d5b284426bf343e3a5eb7958e04d769557c/pkg/ccl/changefeedccl/cdcevent/settings.go#L19)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

we probably don't need to introduce top level settings file and the settings themselves do not need to be exported.
This can be defined in the same file that uses these settings.
</blockquote></details>

Done.
___
*[`pkg/ccl/changefeedccl/cdcevent/settings.go` line 29 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/112342#-NgfCNyb95OCZMAcwNbb:-Nh7GZii5OFnhUxZbhg_:b-cwib81) ([raw file](https://github.com/cockroachdb/cockroach/blob/903a6d5b284426bf343e3a5eb7958e04d769557c/pkg/ccl/changefeedccl/cdcevent/settings.go#L29)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

I see, so we do have a setting for this. I think we need to lift it up into `row` package, and make `0s` the default that means that every occurrence is sampled.
</blockquote></details>

I want to push back a little bit on having a global cluster setting. If you have a global setting, then all row fetcher instances will start logging KVs - sometimes you might only be debugging a changefeed so you want to enable logging for changefeed rows only. For this reason, I put the setting in the changefeed package and implementing the sampling rate as an argument to the row fetcher instance. `TraceKV` is also an argument and not a cluster setting. I suspect for the same reason.
___
*[`pkg/sql/row/fetcher.go` line 1144 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/112342#-NgfBY0z4V5uF6fvQNcR:-Nh7GRUdCFbskaaf3hve:b-11fzck) ([raw file](https://github.com/cockroachdb/cockroach/blob/615ea972162024023d2d2dc7e74d63ed66621a1f/pkg/sql/row/fetcher.go#L1144)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

I think that we need to preserve the default behavior - namely that every KV gets "fetched ..." line printed in the trace by default. We should introduce a cluster setting that would set the "KV sample frequency", which would have 0s by default (meaning everything is logged), and then we could reduce the frequency for the use case you have in mind.
</blockquote></details>

I think preserving the default behavior is good. I changed the default to 0s. When using the row fetcher in changefeeds, there was no logging before and no default behavior to follow. I think sticking with 1s for changefeeds is good. If a node changefeed is emitting 100k messages/sec, there will be a lot of logs produced if we log each one.

**Review by jayshrivastava — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/112342)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner, @miretskiy, and @yuzefovich)

___
*[`pkg/sql/row/fetcher.go` line 1144 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/112342#-NgfBY0z4V5uF6fvQNcR:-Nh7HqaI4d1iEnDKoo06:b-cj3o4g) ([raw file](https://github.com/cockroachdb/cockroach/blob/615ea972162024023d2d2dc7e74d63ed66621a1f/pkg/sql/row/fetcher.go#L1144)):*
<details><summary><i>Previously, jayshrivastava (Jayant) wrote…</i></summary><blockquote>

I think preserving the default behavior is good. I changed the default to 0s. When using the row fetcher in changefeeds, there was no logging before and no default behavior to follow. I think sticking with 1s for changefeeds is good. If a node changefeed is emitting 100k messages/sec, there will be a lot of logs produced if we log each one.
</blockquote></details>

There also might be multiple changefeed processors on the node, each processing 100k rows/sec.

File: pkg/ccl/changefeedccl/cdcevent/settings.go [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+import (
+	"time"
+
+	"github.com/cockroachdb/cockroach/pkg/settings"
+)
+
+// RowFetcherTraceKVs controls if we write the experimental new format BACKUP
+// metadata file.
+var RowFetcherTraceKVs = settings.RegisterBoolSetting(
```
Comment (miretskiy) [inline pkg/ccl/changefeedccl/cdcevent/settings.go:19]: I'm actually wondering if having these settings is the right thing to do.
We are (ab)using these settings in great many places.
These are intended to debug low level failures.
Could the same functionality be accomplished via vlog?
For example, we can say "vmodule=rowfetcher_cache=3" ...
When we return fetcher, we can check `if log.V(3)` ... enable tracing.

Comment (jayshrivastava) [reply pkg/ccl/changefeedccl/cdcevent/settings.go]: I don't feel too strongly about using settings vs `if log.V(3)`. I do think that having a setting for the frequency is a no brainer (it's certainly a problem if we start logging too many rows / sec right?). But TraceKV can be enabled with the log.V(3) bool.

Comment (miretskiy) [reply pkg/ccl/changefeedccl/cdcevent/settings.go]: even frequency doesn't need to be enabled here -- as long as you can set frequence as you do with TraceKV fetcher option.  It's fine to keep it a constant (unless we find that we absolutely need to be able to control it dynamically)

Comment (jayshrivastava) [reply pkg/ccl/changefeedccl/cdcevent/settings.go]: I feel like there's cases where you want to change the frequency. If a changefeed has 300M rows, then limiting to 1s is good. If it only has like 3 rows, then we should print all the KVs so we don't miss anything.

**Review by yuzefovich — APPROVED** [review]
:lgtm:

> Reviewed 1 of 3 files at r3, 2 of 4 files at r4, 3 of 3 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/112342#-:-Nh8_YZMEQm4Kb0WNSYj:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @jayshrivastava, @mgartner, and @miretskiy)

___
*[`pkg/ccl/changefeedccl/cdcevent/rowfetcher_cache.go` line 36 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/112342#-Nh8ZyBk9fe6OBmRs0aJ:-Nh8ZyBk9fe6OBmRs0aK:bqdj6u5) ([raw file](https://github.com/cockroachdb/cockroach/blob/d824910c44d41bc442782a37af46fbc6da62b070/pkg/ccl/changefeedccl/cdcevent/rowfetcher_cache.go#L36)):*
> ```Go
> )
> 
> // RowFetcherTraceKVs enables pretty KV generation in the row fetcher so KVs will
> ```

nit: `s/RowFetcherTraceKVs/traceKVs/`.
___
*[`pkg/ccl/changefeedccl/cdcevent/rowfetcher_cache.go` line 46 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/112342#-Nh8_487DmN9ERSY4Zas:-Nh8_48881orBUpGEQ-l:b-4vk5cl) ([raw file](https://github.com/cockroachdb/cockroach/blob/d824910c44d41bc442782a37af46fbc6da62b070/pkg/ccl/changefeedccl/cdcevent/rowfetcher_cache.go#L46)):*
> ```Go
> )
> 
> // RowFetcherTraceKVLogFrequency controls how frequently KVs are logged when
> ```

nit: `s/RowFetcherTraceKVLogFrequency/traceKVLogFrequency/`.
___
*[`pkg/sql/row/fetcher.go` line 1144 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/112342#-NgfBY0z4V5uF6fvQNcR:-Nh8ZpOQ5W1KnSMl-o48:b-vmrj41) ([raw file](https://github.com/cockroachdb/cockroach/blob/615ea972162024023d2d2dc7e74d63ed66621a1f/pkg/sql/row/fetcher.go#L1144)):*
<details><summary><i>Previously, jayshrivastava (Jayant) wrote…</i></summary><blockquote>

There also might be multiple changefeed processors on the node, each processing 100k rows/sec.
</blockquote></details>

SGTM, thanks.
___
*[`pkg/ccl/changefeedccl/cdcevent/settings.go` line 29 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/112342#-NgfCNyb95OCZMAcwNbb:-Nh8_CkO3jGuhusahetz:bvh6zdm) ([raw file](https://github.com/cockroachdb/cockroach/blob/903a6d5b284426bf343e3a5eb7958e04d769557c/pkg/ccl/changefeedccl/cdcevent/settings.go#L29)):*
<details><summary><i>Previously, jayshrivastava (Jayant) wrote…</i></summary><blockquote>

I want to push back a little bit on having a global cluster setting. If you have a global setting, then all row fetcher instances will start logging KVs - sometimes you might only be debugging a changefeed so you want to enable logging for changefeed rows only. For this reason, I put the setting in the changefeed package and implementing the sampling rate as an argument to the row fetcher instance. `TraceKV` is also an argument and not a cluster setting. I suspect for the same reason.
</blockquote></details>

Makes sense, sounds good to me.

**Comment by jayshrivastava** [comment]
For posterity, using `crdb_internal.set_vmodule('fetcher=2,rowfetcher_cache=2')` is the handy command to use.

**Pull request merged**